info-compiler/src/main/java/com/google/sps/webcrawler/NewsContentProcessor.java

```diff
@@ -15,23 +15,239 @@
+    newsArticle.setAbbreviatedContent(abbreviatedContent);
+    return newsArticle;
+  }
+
+  /**
+   * Extractively summarizes the news article content by computing inter-sentence similarity and
+   * applying PageRank to find the most meaningful sentences. The chosen sentences are arranged
+   * in the original order they appeared in.
+   *
+   * @throws Exception if {@code SentenceModel} or {@code TokenizerModel} model instantiation
+   *     fails, of if {@code PageRank} algorithm fails. For example, {@code SENTENCE_DETECTOR_FILE}
+   *     and {@code TOKENIZER_FILE} must be prepared.
+   * @see <a href="https://towardsdatascience.com/understand-text-summarization-and-create-your-
+   *     own-summarizer-in-python-b26a9f09fc70></a>
+   */
+  public static NewsArticle summarize(NewsArticle originalNewsArticle) throws Exception {
+    NewsArticle newsArticle = new NewsArticle(originalNewsArticle);
+    String rawContent = newsArticle.getContent();
+    String[] sentences = breakIntoSentences(rawContent);
+    if (sentences.length <= SUMMARIZATION_MAX_SENTENCE_NUMBER) {
+      newsArticle.setSummarizedContent(rawContent);
+      return newsArticle;
+    }
+    Graph<IntValue, StringValue, DoubleValue> similarityGraph = buildSimilarityGraph(sentences);
+    List<PageRank.Result<IntValue>> ranking = getRanking(similarityGraph);
+    String summarizedContent = extractSentencesBasedOnRanking(ranking, sentences);
+    newsArticle.setSummarizedContent(summarizedContent);
     return newsArticle;
   }
+
+  /** Breaks down {@code rawContent} into sentences. */
+  private static String[] breakIntoSentences(String rawContent) throws IOException {
+    InputStream modelFile = new FileInputStream(SENTENCE_DETECTOR_FILE);
+    SentenceModel model = new SentenceModel(modelFile);
+    SentenceDetectorME sentenceDetector = new SentenceDetectorME(model);
+    return sentenceDetector.sentDetect(rawContent);
+  }
+
+  /**
+   * Builds the similarity graph among all sentences. The vertex (k, vv) represents the k-th
+   * sentence. The edge (k1, k2, ev) represents a similarity of ev between the k1-th and k2-th
+   * sentences. An edge exists only when the two sentences are similar "enough", as determined
+   * by {@code SIMILARITY_THRESHOLD}.
+   */
+  private static Graph<IntValue, StringValue, DoubleValue> buildSimilarityGraph(String[] sentences)
+      throws IOException {
+    List<Vertex<IntValue, StringValue>> vertices =
+        new ArrayList<Vertex<IntValue, StringValue>>(sentences.length);
+    List<Edge<IntValue, DoubleValue>> edges =
+        new ArrayList<Edge<IntValue, DoubleValue>>((int) Math.pow(sentences.length, 2));
+    for (int i = 0; i < sentences.length; i++) {
+      vertices.add(new Vertex(new IntValue(i), new StringValue(sentences[i])));
+      for (int j = i; j < sentences.length; j++) {
+        if (i == j) {
+          edges.add(new Edge(new IntValue(i), new IntValue(j), new DoubleValue(0)));
+          continue;
+        }
```

Comment:
What's the value for adding a 0 edge between a sentence and itself?

Reply:
Thanks for the reminder! There shouldn't an edge between a sentence and itself. I'll fix this

Reply:
Technically, the vertex values and edge values don't matter. Only the existence of edges between vertices matters.